### PR TITLE
Source serialization design

### PR DESCRIPTION
**Pull request by davidbau**
Hi folks, this is a large (vibe coded) code change to start the discussion; I'm happy to discuss it. It's pretty exhaustively unit tested, but I will need help reviewing and landing it.

This PR introduces **source-based serialization** as an alternative to cloudpickle for remote execution on NDIF servers. The key addition is the `@nnsight.remote` decorator that marks functions and classes as safe for remote execution.

  **Why to do this:**
  - **Cross-version compatibility**: Clients on Python 3.10 can work with servers on Python 3.12+ (previously cloudpickle serialized version-specific bytecode)
  - **Organic library ecosystem**: Third-party libraries like nnterp can work with NDIF without server-side installation
  - **Fail-fast validation**: Incompatible code errors at import time with clear messages, instead of runtime failures

  **Backward compatible**: Existing cloudpickle-based clients continue to work (with a deprecation warning).  We should eventually remove this codepath.

  **Visualize the code for this change**: https://davidbau.github.io/nnsight/callgraph.html#c=38b9c2f - for the 12 affected files.
  **Narrative discussion**: https://davidbau.github.io/nnsight/source-serialization-tutorial.html
  **Technical design**: https://davidbau.github.io/nnsight/nnsight-source-serialization-design.html
  
  **Example usage:**
  ```python
  import json
  import torch
  import torch.nn as nn
  import torch.nn.functional as F

  # Custom user class and data
  class SentimentProbe(nn.Module):
      def __init__(self, hidden_size):
          super().__init__()
          self.linear = nn.Linear(hidden_size, 2)
      def forward(self, x):
          return self.linear(x)
   with open("sentiment_data.json") as f:
      training_data = json.load(f)  # [{"text": "...", "label": 0}, ...]
  probe = SentimentProbe(768)
  optimizer = torch.optim.Adam(probe.parameters())

  # Code dependencies are discovered and sent to the server so "probe" can be trained.
  with model.session(remote=True):
      for epoch in range(10):
          for example in training_data:
              with model.trace(example["text"]):
                  hidden = model.transformer.h[10].output[0][:, -1, :]
                  logits = probe(hidden)
                  loss = F.cross_entropy(logits, torch.tensor([example["label"]]))
                  loss.backward()
              optimizer.step()
              optimizer.zero_grad()

      trained_weights = probe.state_dict().save()
  ```
  **Key Features**

  - @nnsight.remote decorator (src/nnsight/remote.py) - validates and extracts source code for functions and classes
  - Comprehensive serialization (src/nnsight/intervention/serialization_source.py):
    - Functions, classes, lambdas with closure capture
    - Tensor serialization (Base64 with compression)
    - nn.Module subclasses with recursive state serialization
    - Cycle detection and object deduplication via __id__/__ref__
  - Security layer (src/nnsight/intervention/restricted_execution.py) - RestrictedPython integration with audit logging
  - Error tracebacks - preserves original file/line numbers in remote execution errors

  **Changes**

  - src/nnsight/remote.py - new `@nnsight.remote` decorator with import-time validation
  - src/nnsight/intervention/serialization_source.py - source-based serialization engine
  - src/nnsight/intervention/restricted_execution.py - RestrictedPython security guards
  - src/nnsight/intervention/serialization.py - integration with existing serialization
  - src/nnsight/schema/request.py - schema updates for source-based payloads
  - docs/nnsight-source-serialization-design.md - detailed design document

 **Testing**

  Added 199 new tests covering the `@nnsight.remote` decorator, serialization edge cases, and end-to-end scenarios:
  - tests/test_remote.py - decorator validation, serialization, lambdas, tensors
  - tests/test_serialization_edge_cases.py - slots, metaclasses, circular refs, inheritance
  - tests/test_e2e_remote_nnterp.py - end-to-end integration tests

  Key scenarios tested:
  - Training a custom probe class remotely
  - Using nnterp.StandardizedTransformer in remote traces
  - nn.Module subclass roundtrip serialization

  Test plan

  - All 273 tests pass locally
  - Existing test_lm.py and test_tiny.py tests unaffected
  - Help needed: Manual testing on a development NDIF server

### CODE REVIEW

File: src/nnsight/intervention/serialization_source.py [hunk 1]
```diff
@@ -0,0 +1,2791 @@
...
+        "Convert to a dense tensor:\n"
+        "  tensor = torch.tensor(sparse_matrix.toarray())\n"
+        "Or extract the CSR components if you need sparse format."
+    ),
+    'scipy.sparse._csc.csc_matrix': (
+        "scipy.sparse.csc_matrix cannot be serialized for remote execution.\n"
+        "\n"
+        "Convert to a dense tensor:\n"
+        "  tensor = torch.tensor(sparse_matrix.toarray())"
+    ),
```
Comment (Butanium) [inline src/nnsight/intervention/serialization_source.py]: This looks very brittle and prone to errors. I had a similar PR generated by claude code using this kind of hacks and it broke some stuff that it didn't think about
